### DISCUSSION TITLE
Imperative style support.

### DISCUSSION THREAD

**Discussion by mr-mapache** [sqlalchemy/sqlalchemy#11793] [Ideas]
Let's say I have a domain model with a lot of stuff that cannot or shouldn't be persisted in database, like for example User aggregate root, with a Deque[Event] member and a publisher, also a List[Sessions] that are stored in redis, stuff like that, and thedatabase don't fit the domain layer. Declarative base is just usless in this case and you have to use imperative style. 

Using the declarative base as imperative just don't fix the issue, suppose you have a users table in a database with the following style:

```python
users = Table(
    'users',
    metadata,
    Column('id', Integer, primary_key=True), # surrogate key, don't have anything to do with the domain id.
    Column('user_id', UUID, unique=True),
    Column('user_username', String, unique=True),
)
```

This would be the declarative style:

```python
class Schema(DeclarativeBase):
    surrogate_key: Mapped[int] = mapped_column('id', primary_key=True) #have to rename this key from database.

class User(Schema):
    __tablename__ = 'users'
    id: Mapped[UUID] = mapped_column('user_id', unique=True, nullable=False)
    username: Mapped[str] = mapped_column('user_username', unique=True, nullable=False)
```

As you see, in the first one, it's very clear I'm talking about a table that shouldn't be initialized, that is an infrastructure concern and has to be used with the insert, select, update, delete style. But in the second one example, is very easy for another person to confuse the User "table" with the User aggregate, cannot import those two without an alias as their names collides. 

I tried an easy solution for having PEP support was to mimic imperative style with declarative, but I think is just code smell and confusing, so I went back to use the classic imperative style with tuples and string column names in my data access objects. 

```python
class users(Schema):
    __tablename__ = 'users'
    id: Mapped[int] = mapped_column('id', primary_key=True)
    user_id: Mapped[UUID] = mapped_column('user_id', unique=True)
    user_username: Mapped[str] = mapped_column('user_username', unique=True, nullable=False)
```
Shouldn't be to hard to create something like a Table factory returning a dynamically created schema, I saw people doing this kind of stuff for creating tables dynamically to address the problem. 

```python

tests = type('tests', (DeclarativeBase,), {
    '__tablename__': 'tests',
    'test_id': Column(Integer, primary_key=True, autoincrement=True),
    'fldA': Column(String),  
    ... other columns
    }
)
```

There are a lot of post over the internet complaining about the lack of support for imperative style. Some people creating their own data mappers, people like this [guy](https://www.reddit.com/r/Python/comments/rt3apo/developing_a_python_database_orm_with_a_different/) trying to create his own ORM because of this, even the book "architecture patterns with python"  o'reilly promotes the imperative style. 

The truth is that for simple CRUD apps, where you have full controll over your data, it's nice to have declarative, but when you have a domain model or you are developing with an existant database it's just become a mess to use declarative, with a lot of DTO's, factories and even some dirty hacks.

**Comment by CaselIT** [comment]
Hi,

I think there's a bit of confusion here. Table is not what imperative style is.

Imperative style referers the way of mapping the tables and classes described here https://docs.sqlalchemy.org/en/20/orm/mapping_styles.html#imperative-mapping . The `map_imperatively` call there.

Also note that while it was originally the plan to deprecate it, that olan was abandoned since there are a lot of users that like it, that's no longer the plan. There were some internal changes that made the various mapping styles supported by the orm all use the same code, so the maintenance effort has been reduced 

The table definition you have there is part of core, and without table pretty much nothing in sqlalchemy would work. Your user example generates a table too: that's placed in the `__table__` attribute of the class. See https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.DeclarativeBase.__table__

**Reply by mr-mapache** [reply]
Thanks, really like the table stuff I didn't want it to be deprecated, I didn't find anything like that for python other than raw sql. Anyway would be nice to se some work on it some day, in things like I mentioned.

Maybe a wrapper that has field names instead of a dict in the .columns property in the table, or when making queries having a return type that has fields and not tuples (I know that rows in a database are tuples but it's a bit confusing somethimes).

For example this:

`
command = delete(users).where(users.columns['user_id'] == user.id)
`

Could be this:
`
command = delete(users).where(users.user_id] == user.id)
`

Or this:

`
user = user_factory.create(row[1], row[2])
`

Could be this:

`
user = user_factory.create(row.user_id, row.user_name)
`

**Reply by CaselIT** [reply]
It already works that way. Just use .c to access the column `command = delete(users).where(users.c.user_id] == user.id)`

(You can access .columns too by attribute)

**Reply by mr-mapache** [reply]
Yeah, the .c accessor doesn't have intellisense support, since it returns a ColumnCollection created in runtime and do not access the global name scopes. So you type .c. and don't get any fields like with declarative style. 

I've looking throught the code of sqlalchemy and I don't think that tere is an easy work around for this, inserting dynamically created names in the naming scope is not a good idea. Anyway, fixing that won't fix the fact that when you do a query it returns a tuple, not a class with variables. 

I think I will stick with the declarative style and make a type alias for using it like a global table. Thanks anyway.

**Reply by CaselIT** [reply]
> Yeah, the .c accessor doesn't have intellisense support, since it returns a ColumnCollection created in runtime and do not access the global name scopes. So you type .c. and don't get any fields like with declarative style.

> I've looking throught the code of sqlalchemy and I don't think that tere is an easy work around for this, inserting dynamically created names in the naming scope is not a good idea. Anyway, fixing that won't fix the fact that when you do a query it returns a tuple, not a class with variables.

it's not possible to show intellisense with the current status of typing in python, since the properties need to be statically defined. if you want intellisense use declarative or do your own wrapper wrapper kind, something like: -- see my other comment


>  Anyway, fixing that won't fix the fact that when you do a query it returns a tuple, not a class with variables.

like I mentioned above, it does work like you expect. try it. The returned object acts like a named tuple

**Comment by CaselIT** [comment]
An idea to enable type checkers on table would be to trick the type checker a bit:
```py
from typing import Any, Generic, TypeVar
import sqlalchemy as sa

meta = sa.MetaData()

T = TypeVar("T")

class TypedTable(sa.Table, Generic[T]):
    c: T  # type: ignore
    column: T

def typed_table(name: str, meta: sa.MetaData, cls: T, **kwargs: Any) -> TypedTable[T]:

    cols = []
    for name, col in vars(cls).items():
        if isinstance(col, sa.Column):
            if col.name is None:
                col.name = name
            cols.append(col)

    return sa.Table(name, meta, *cols, **kwargs)  # type: ignore

class my_tbl_cols:
    id = sa.Column(sa.Integer, primary_key=True)
    name = sa.Column(sa.String)
    age = sa.Column(sa.Integer, nullable=False)

my_tcl = typed_table("my_tbl", meta, my_tbl_cols)

print(repr(my_tcl))
```

@zzzeek that do you think about having an utility constructor?

**Reply by CaselIT** [reply]
I liked the idea and implemented a version that doesn't lie to the type checker: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5456

a bit of polishing is needed, but it works fine.

@zzzeek not sure if this lie to the type checker version better. what do you think? It would avoid having another property, but it would trip up when using columns that are named like the attributes of the column collection

**Comment by zzzeek** [comment]
> I saw in a discussion that you are abandoning the imperative style support. I think you are making a mistake.

This is not correct, there are no plans to abandon imperative mappings and they remain an essential feature of the SQLAlchemy ORM.    Imperative mappings are prominently featured throughout SQLAlchemy's ORM documentation and as you can see are a top level item in our main ORM configuration documentation at https://docs.sqlalchemy.org/en/20/orm/mapping_styles.html#orm-mapping-styles .   

If you want to link to that discussion and/or clarify that whoever made that claim is incorrect, thanks.

> I would like to know if you are abandoning the imperative style to just stop using and promoting the usage of this library.

if you're going to make charged comments like this, I would ask if you at least read the SQLAlchemy tutorial before running down this path?   I really dont know where this notion that "imperative style is being abandoned" came from, it is all over the documentation in big bold sections, no deprecation messages, it's not marked as "legacy", when we "abandon" something, it's very obvious.  I am really confused

https://docs.sqlalchemy.org/en/20/tutorial/metadata.html#working-with-database-metadata

**Reply by mr-mapache** [reply]
Sorry, I saw your comment: 

> I was hoping to remove imperative mapping from 2.0

But missinterpreted and overreacted a bit, I though imperative mapping was refering to table style and while still in use in the library, just abandoned without new features.

**Reply by zzzeek** [reply]
can you link to where that is?  that was a long time ago so i want to edit that and make sure it's not misleading

**Reply by mr-mapache** [reply]
> can you link to where that is? that was a long time ago so i want to edit that and make sure it's not misleading
[here](
https://github.com/sqlalchemy/sqlalchemy/discussions/9180#discussioncomment-10470787)

**Comment by mr-mapache** [comment]
Thanks for support guys, and sorry for being such an asshole, I wasn't having a good day. Let me know if you need something or if I can help with anything.
